### PR TITLE
Pointer arithmetic affects the alignment factor

### LINKED ISSUE #1528: pointer math fails to modify alignment of type
```zig
const std = @import("std");

test "aoeu" {
    var arr: [10]u8 align(2) = undefined;
    var x: usize = 1;

    const ptr = ([*]u8)(&arr);
    @compileLog(@typeOf(ptr));
    const ptr2 = ptr + 1;
    @compileLog(@typeOf(ptr2));
    const ptr3 = ptr + x;
    @compileLog(@typeOf(ptr3));
}
```

expected output:

```
| [*]align(2) u8
| [*]align(1) u8
| [*]align(1) u8
```

actual output:

```
| [*]align(2) u8
| [*]align(2) u8
| [*]align(2) u8
```

When we add a comptime known value to the pointer, we know what the new alignment should be.

When we add a runtime value to the pointer, it should make the type have alignment 1 no matter what, because the value could have made it unaligned.

### PR DESCRIPTION
**Pull request by LemonBoy**
Closes #1528

### CODE REVIEW

**Comment by LemonBoy** [comment]
> This case regressed:

Hm? You're increasing `ptr` address by `4` (since it's an array of `u32`) and that makes it not aligned to 8 anymore

**Comment by andrewrk** [comment]
Yes but it should still be aligned to 4, e.g. it should not be underaligned. Maybe this is a better example:

```zig
test "extra aligned ptr" {
    var x: [2]u32 align(8) = [_]u32{1, 2};

    const result = addTwoNumbers(&x);
}

fn addTwoNumbers(ptr: [*]align(8) u32) u32 {
    return fetch(ptr + 0) + fetch(ptr + 1);
}

fn fetch(ptr: [*]u32) u32 {
    return ptr[0];
}
```

This correctly passes in master branch.

**Comment by LemonBoy** [comment]
> Yes but it should still be aligned to 4, e.g. it should not be underaligned.

Gotcha, I've added some code to do so ;)

File: src/ir.cpp [hunk 1]
```diff
@@ -15776,19 +15776,51 @@ static IrInstruction *ir_analyze_bin_op_math(IrAnalyze *ira, IrInstructionBinOp
...
+        if (op2_val != nullptr) {
+            uint32_t align_bytes;
+            if ((err = resolve_ptr_align(ira, op1->value->type, &align_bytes)))
+                return ira->codegen->invalid_instruction;
+
+            if (byte_offset & (align_bytes - 1)) {
+                // The resulting pointer is aligned to the lcd between the
+                // offset (an arbitrary number) and the alignment factor (always
+                // a power of two, non zero)
+                uint32_t new_align = 1 << ctzll(byte_offset | align_bytes);
```
Comment (andrewrk) [inline src/ir.cpp]: :clap: :clap: :clap: 

much sophisticated, very wow :dog:

File: test/stage1/behavior/pointers.zig [hunk 2]
```diff
@@ -288,3 +288,26 @@ test "pointer to array at fixed address" {
...
+
+        expect(@typeInfo(@TypeOf(ptr)).Pointer.alignment == 8);
+        const ptr1 = ptr + 1; // 1 * 4 = 4 -> lcd(4,8) = 4
+        expect(@typeInfo(@TypeOf(ptr1)).Pointer.alignment == 4);
+        const ptr2 = ptr + 4; // 4 * 4 = 16 -> lcd(16,8) = 8
+        expect(@typeInfo(@TypeOf(ptr2)).Pointer.alignment == 8);
+        const ptr3 = ptr + 0; // no-op
+        expect(@typeInfo(@TypeOf(ptr3)).Pointer.alignment == 8);
+        const ptr4 = ptr + x; // runtime-known addend
+        expect(@typeInfo(@TypeOf(ptr4)).Pointer.alignment == 1);
```
Comment (andrewrk) [inline test/stage1/behavior/pointers.zig]: Wait, this should still be aligned to u32. `x` is runtime known but the number of bytes the pointer marches forward is `@sizeOf(u32) * x` which is guaranteed to be divisible by `@sizeOf(u32)`.

Comment (LemonBoy) [reply test/stage1/behavior/pointers.zig]: Yeah, I focused too much on the comptime-known branch and forgot the other one, now it should be fine!

**Pull request merged**